File: frontend/src/types/governance.ts
/* -------------------------------------------------------------------------- */
/*  Governance validation types                                               */
/* -------------------------------------------------------------------------- */

export const GovernanceSeverity = {
  ERROR: 'ERROR',
  WARNING: 'WARNING',
  INFO: 'INFO',
} as const;

export type GovernanceSeverity =
  (typeof GovernanceSeverity)[keyof typeof GovernanceSeverity];

export interface Violation {
  /** Machine-readable rule code (e.g. "GOV-001"). */
  ruleId: string;
  /** Human-readable rule name. */
  ruleName: string;
  /** Severity of this violation. */
  severity: GovernanceSeverity;
  /** Human-readable description of the violation. */
  message: string;
  /** JSON path or field that triggered the violation. */
  field?: string;
  /** Suggestion on how to fix the violation. */
  remediation?: string;
}

export interface GovernanceResult {
  /** Whether all ERROR-severity rules passed. */
  passed: boolean;
  /** All violations found (may include WARNINGs even when passed is true). */
  violations: Violation[];
  /** Total rules evaluated. */
  rulesEvaluated: number;
  /** ISO timestamp of the evaluation. */
  evaluatedAt: string;
}

export interface GovernanceRule {
  /** Unique rule identifier. */
  ruleId: string;
  /** Human-readable name. */
  name: string;
  /** Detailed description of the rule. */
  description: string;
  /** Severity if violated. */
  severity: GovernanceSeverity;
  /** Whether this rule is currently enabled. */
  enabled: boolean;
  /** Category grouping (e.g. "naming", "security", "capacity"). */
  category: string;
}

export interface GovernanceValidateRequest {
  onboardingId?: string;
  /** Inline data to validate (alternative to onboardingId). */
  data?: Record<string, unknown>;
}

File: backend/app/schemas/governance.py
"""Pydantic schemas for the governance rules engine."""

from __future__ import annotations

from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import GovernanceSeverity


class GovernanceValidateRequest(BaseModel):
    """Request to run governance validation on an onboarding submission."""

    model_config = ConfigDict(str_strip_whitespace=True)

    onboarding_request_id: UUID
    dry_run: bool = Field(
        default=False,
        description="When True, violations are returned but the request is not blocked.",
    )


class Violation(BaseModel):
    """A single governance rule violation."""

    rule_id: str = Field(
        description="Unique identifier for the governance rule.",
        examples=["GOV-001"],
    )
    severity: GovernanceSeverity
    message: str = Field(description="Human-readable explanation of the violation.")
    field: str | None = Field(
        default=None,
        description="Dot-path of the offending field, if applicable.",
        examples=["telemetry_scope.metrics.enabled"],
    )
    suggestion: str | None = Field(
        default=None,
        description="Actionable suggestion to resolve the violation.",
    )
    metadata: dict[str, Any] = Field(default_factory=dict)


class GovernanceResult(BaseModel):
    """Aggregate governance validation result."""

    model_config = ConfigDict(from_attributes=True)

    onboarding_request_id: UUID
    passed: bool = Field(
        description="True if no hard violations were found.",
    )
    score: float = Field(
        ge=0.0,
        le=100.0,
        description="Governance compliance score (100 = fully compliant).",
    )
    hard_violations: list[Violation] = Field(
        default_factory=list,
        description="Blocking violations that must be fixed before submission.",
    )
    soft_violations: list[Violation] = Field(
        default_factory=list,
        description="Advisory violations (warnings) that do not block submission.",
    )
    info_notices: list[Violation] = Field(
        default_factory=list,
        description="Informational notices for awareness.",
    )
    evaluated_rules_count: int = Field(
        ge=0,
        description="Total number of rules evaluated.",
    )
